packages/grafana-ui/src/components/DataSourceSettings/CustomHeadersSettings.tsx

```diff
@@ -112,27 +118,25 @@ export class CustomHeadersSettings extends PureComponent<Props, State> {
 
   updateSettings = () => {
     const { headers } = this.state;
-    const { jsonData } = this.props.dataSourceConfig;
-    const secureJsonData = this.props.dataSourceConfig.secureJsonData || {};
+    const newSecureJsonData = { ...(this.props.dataSourceConfig.secureJsonData || {}) };
+
+    // we remove every httpHeaderName* field first
+    const newJsonData = Object.fromEntries(
+      Object.entries(this.props.dataSourceConfig.jsonData).filter(([key, val]) => !key.startsWith('httpHeaderName'))
+    );
+
+    // then we add the current httpHeader-fields
     for (const [index, header] of headers.entries()) {
-      jsonData[`httpHeaderName${index + 1}`] = header.name;
+      newJsonData[`httpHeaderName${index + 1}`] = header.name;
       if (!header.configured) {
-        secureJsonData[`httpHeaderValue${index + 1}`] = header.value;
+        newSecureJsonData[`httpHeaderValue${index + 1}`] = header.value;
```

Comment:
Took me a while to figure out what this code did before. Another problem that is still there is that we do not remove the value from secureJsonData can we also add that? Cause right now you can configure a header then delete it, configure another header without setting the value and it will inherit the previous orphaned value which isn't great.

Reply:
Good catch!

Reply:
done